File: core/lib/Models/Voucher.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using Innofactor.EfCoreJsonValueConverter;
using our.orders.Helpers;
using our.orders.Repositories;

namespace our.orders.Models
{
    public class Voucher : Model
    {
        public decimal InitialValue { get; set; }
        public decimal Value { get; set; }
        public string Code { get; set; }
        public string Currency { get; set; }
        public DateTime? Expiration { get; set; }
        public bool Used { get; set; }
        public bool MultipleUse { get; set; }


        [JsonField]
        public IEnumerable<string> OrderIds  { get; set; }

        public bool Valid { get => !Expired && !Used; set { } }

        public bool Expired { get => Expiration.HasValue && Expiration.Value.CompareTo(DateTime.UtcNow) < 0; set { } }
        public override string Preview() => Code;
    }
}

File: core/app/src/@types/our-orders/Voucher.ts
import { ModelBase } from '.';

export interface Voucher extends ModelBase {
   InitialValue: number;
   Value: number;
   Code: string;
   Currency: string;
   Expiration: Date | string | null;
   MultipleUse: boolean;
   
   Used: boolean;

   Expired: boolean;

   Valid: boolean;
   OrderIds: string[];
   Preview: string;
}
